packages/frontend/src/features/sqlRunner/components/Header/HeaderEdit.tsx

```diff
@@ -41,6 +43,37 @@ export const HeaderEdit: FC = () => {
     const config = useAppSelector((state) =>
         selectChartConfigByKind(state, selectedChartType),
     );
+
+    // Store initial chart config to detect if there are any changes later on
+    const [initialSavedSqlChart, setInitialSavedSqlChart] = useState<
+        SqlChart | undefined
+    >();
+    const [initialChartConfig, setInitialChartConfig] = useState<
+        | VizBarChartConfig
+        | VizLineChartConfig
+        | VizPieChartConfig
+        | VizTableConfig
+        | undefined
+    >();
+
+    useEffect(() => {
+        if (savedSqlChart && initialSavedSqlChart === undefined)
+            setInitialSavedSqlChart(savedSqlChart);
+    }, [savedSqlChart, initialSavedSqlChart]);
+
+    useEffect(() => {
+        if (config && initialChartConfig === undefined)
+            setInitialChartConfig(config);
+    }, [config, initialChartConfig]);
+
+    const hasChanges = useCallback(() => {
+        if (!initialSavedSqlChart || !initialChartConfig) return false;
+        return (
+            sql !== initialSavedSqlChart.sql ||
+            limit !== initialSavedSqlChart.limit ||
+            JSON.stringify(config) !== JSON.stringify(initialChartConfig)
+        );
+    }, [initialSavedSqlChart, initialChartConfig, sql, limit, config]);
```

Comment:
Let's make `hasChanges` a `useMemo` variable so it matches the intent of this value

```tsx
const hasChanges = useMemo(() => {
        if (!initialSavedSqlChart || !initialChartConfig) return false;
        const changedSql = sql !== initialSavedSqlChart.sql;
        const changedLimit = limit !== initialSavedSqlChart.limit;
        const changedConfig = !isEqual(config, initialChartConfig);

        return changedSql || changedLimit || changedConfig;
    }, [initialSavedSqlChart, initialChartConfig, sql, limit, config]);
```

replaced with `isEqual` bc we might reorder some keys and json stringify comparison will pick up on _false positives_